kanidmd/src/lib/idm/oauth2.rs

```diff
@@ -0,0 +1,1035 @@
+//! Oauth2 resource server configurations
+//!
+//! This contains the in memory and loaded set of active oauth2 resource server
+//! integrations, which are then able to be used an accessed from the IDM layer
+//! for operations involving oauth2 authentication processing.
+//!
+
+use crate::identity::IdentityId;
+use crate::prelude::*;
+use concread::cowcell::*;
+use fernet::Fernet;
+use hashbrown::HashMap;
+use kanidm_proto::v1::UserAuthToken;
+use openssl::sha;
+use time::OffsetDateTime;
+use url::{Origin, Url};
+use webauthn_rs::base64_data::Base64UrlSafeData;
+
+pub use kanidm_proto::oauth2::{
+    AccessTokenRequest, AccessTokenResponse, AuthorisationRequest, CodeChallengeMethod,
+    ConsentRequest, ErrorResponse,
+};
+
+use std::convert::TryFrom;
+use std::time::Duration;
+
+lazy_static! {
+    static ref CLASS_OAUTH2: PartialValue = PartialValue::new_class("oauth2_resource_server");
+    static ref CLASS_OAUTH2_BASIC: PartialValue =
+        PartialValue::new_class("oauth2_resource_server_basic");
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+#[serde(rename_all = "lowercase")]
+pub enum Oauth2Error {
+    // Non-standard
+    AuthenticationRequired,
+    // Standard
+    InvalidRequest,
+    UnauthorizedClient,
+    AccessDenied,
+    UnsupportedResponseType,
+    InvalidScope,
+    ServerError(OperationError),
+    TemporarilyUnavailable,
+}
+
+impl std::fmt::Display for Oauth2Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{}",
+            match self {
+                Oauth2Error::AuthenticationRequired => "authentication_required",
+                Oauth2Error::InvalidRequest => "invalid_request",
+                Oauth2Error::UnauthorizedClient => "unauthorized_client",
+                Oauth2Error::AccessDenied => "access_denied",
+                Oauth2Error::UnsupportedResponseType => "unsupported_response_type",
+                Oauth2Error::InvalidScope => "invalid_scope",
+                Oauth2Error::ServerError(_) => "server_error",
+                Oauth2Error::TemporarilyUnavailable => "temporarily_unavailable",
+            }
+        )
+    }
+}
+
+// == internal state formats that we encrypt and send.
+
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+struct ConsentToken {
```

Comment:
In an oauth2 flow the resource server redirects the users browser to the identity provider to determine if they are authorised to access this resource. If they are authenticated and authorised, the user us requested to provide their consent to proceed to access the resource server, in case it may allow access they do not wish for. If they accept the consent request, they yield to us the consent token that describes the state of the authorisation flow and allows us to proceed.